### PR TITLE
Make JapaneseTokenizer thread-safe and much more efficient.

### PR DESCRIPTION
**Pull request by wmeddie**
## What changes were proposed in this pull request?

JapaneseTokenizer implementation creates new instances of the kuromoji Tokenizer which load the entire dictionary on startup.  By caching the Tokenizer inside the JapaneseTokenizerFactory it becomes much more efficient.  A side effect is that the JapaneseTokenizerFactory also becomes thread-safe.  In Kuromoji only a single thread can create a Tokenizer at a time the tokenizer method is thread-safe.

I also added support for tokenizing into base forms. (A form of stemming.)

## How was this patch tested?

I added tests for the base form functionality and a test that uses threads.

### CODE REVIEW

File: deeplearning4j-nlp-parent/deeplearning4j-nlp-japanese/src/main/java/org/deeplearning4j/text/tokenization/tokenizer/JapaneseTokenizer.java [hunk 1]
```diff
@@ -39,52 +35,57 @@
...
         }
-        return this.preProcess != null ? this.preProcess.preProcess(tokenIter.next()) : tokenIter.next();
+        return getToken(currentToken++);
     }
 
     @Override
     public List<String> getTokens() {
-        return tokens;
+        ArrayList<String> ret = new ArrayList<>(this.tokenCount);
+
```
Comment (agibsonccc) [inline deeplearning4j-nlp-parent/deeplearning4j-nlp-japanese/src/main/java/org/deeplearning4j/text/tokenization/tokenizer/JapaneseTokenizer.java]: Coding best practices:
Interface i = new Implementation();

File: deeplearning4j-nlp-parent/deeplearning4j-nlp-japanese/src/main/java/org/deeplearning4j/text/tokenization/tokenizerfactory/JapaneseTokenizerFactory.java [hunk 2]
```diff
@@ -1,23 +1,45 @@
...
 
     @Override
     public Tokenizer create(String toTokenize) {
         if (toTokenize.isEmpty()) {
             throw new IllegalArgumentException("Unable to proceed; no sentence to tokenize");
         }
-        JapaneseTokenizer t = new JapaneseTokenizer(toTokenize);
+
+        JapaneseTokenizer t = new JapaneseTokenizer(kuromoji, toTokenize, useBaseForm);
+
```
Comment (agibsonccc) [inline deeplearning4j-nlp-parent/deeplearning4j-nlp-japanese/src/main/java/org/deeplearning4j/text/tokenization/tokenizerfactory/JapaneseTokenizerFactory.java]: JapaneseTokenizer t -> Tokenizer t = new JapaneseTokenizer

File: deeplearning4j-nlp-parent/deeplearning4j-nlp-japanese/src/main/java/org/deeplearning4j/text/tokenization/tokenizerfactory/JapaneseTokenizerFactory.java [hunk 3]
```diff
@@ -1,23 +1,45 @@
 package org.deeplearning4j.text.tokenization.tokenizerfactory;
 
+import org.deeplearning4j.text.tokenization.tokenizer.JapaneseTokenizer;
 import org.deeplearning4j.text.tokenization.tokenizer.TokenPreProcess;
 import org.deeplearning4j.text.tokenization.tokenizer.Tokenizer;
-import org.deeplearning4j.text.tokenization.tokenizer.JapaneseTokenizer;
 
 import java.io.InputStream;
```
Comment (agibsonccc) [inline deeplearning4j-nlp-parent/deeplearning4j-nlp-japanese/src/main/java/org/deeplearning4j/text/tokenization/tokenizerfactory/JapaneseTokenizerFactory.java]: Javadoc mentioning the thread safeness and how the kuromoji instance works would make sense here.

File: deeplearning4j-nlp-parent/deeplearning4j-nlp-japanese/src/main/java/org/deeplearning4j/text/tokenization/tokenizerfactory/JapaneseTokenizerFactory.java (hunk 2 above)
Comment (wmeddie) [inline deeplearning4j-nlp-parent/deeplearning4j-nlp-japanese/src/main/java/org/deeplearning4j/text/tokenization/tokenizerfactory/JapaneseTokenizerFactory.java]: All good points. Updates soon.

**Pull request merged**